text/0087-data-types-extensibility.md

```diff
@@ -0,0 +1,130 @@
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+This feature will bring additional compatibility layer in a form of an `enum` wrapper that ties the current (aka legacy) version of the data type and all the updated versions, like so:
+```rust
+pub enum BlockBody {
+    V1(BlockBodyV1),
+    V2(BlockBodyV2),
+    V3(BlockBodyV3),
+    ...
+}
+```
+
+There is no migration of existing data. If a body is stored in the DB as `BlockBodyV1` it'll stay intact and the components that operate on DB will be modified to understand all variants.
+
+It is allowed to used the specific, most recent version directly in the code, because the idea is that only single (most recent) version of a particular type is considered "current". Only parts of the system that need to interact with legacy versions should use the `enum` wrapper, for example, the block synchronizer which must be able to work with historical blocks. This approach will not be enforced, therefore it is possible for all components to work with the `enum` wrapper and react as necessary to handle different variants of data.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+In technical detail, this proposal introduces a new `enum` wrapper type for every data struct we need to enable the extensibility for. For example, this is how the new `Block` type might be represented:
+
+```rust
+pub enum Block {
+    V1(BlockV1),
+    V2(BlockV2),
+}
+
+pub struct BlockV1 {
+    pub(super) hash: BlockHash,
+    pub(super) header: BlockHeader,
+    pub(super) body: BlockBodyV1,
+}
+
+pub struct BlockV2 {
+    pub(super) hash: BlockHash,
+    pub(super) header: BlockHeader,
+    pub(super) body: BlockBodyV2,
+    pub(super) new_member: bool,
+```
+
+Each specific variant is (de)serializable and there is also a (de)serialization implemented for the top level `enum` wrapper which uses the predefined `TAG`s, for example:
+
+```rust
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut buffer = bytesrepr::allocate_buffer(self)?;
+        match self {
+            Block::V1(v1) => {
+                buffer.insert(0, BLOCK_V1_TAG);
+                buffer.extend(v1.to_bytes()?);
+            }
+            Block::V2(v2) => {
+                buffer.insert(0, BLOCK_V2_TAG);
+                buffer.extend(v2.to_bytes()?);
+            }
+        }
+        Ok(buffer)
+    }
+```
+
+Components that never need to touch the legacy types should directly use the most recent version of the datatype. For example, going forward, we never need to gossip "legacy" blocks during the normal chain operation, therefore the `Gossiper` should work directly with `BlockV2`.
+
+The `Fetcher` on the other hand must be able to deal with all variants of the datatype, because it must be able to fetch and understand all historical blocks up to genesis.
+
+## Storage
```

Comment:
nit: I think this should have heading 3 `###`

Reply:
Headings fixed in https://github.com/casper-network/ceps/pull/87/commits/a1e164f0a5f7c0ad1a170f6f112ba5759b6eb6de